torch/utils/data/datapipes/iter/grouping.py

```diff
@@ -27,15 +27,32 @@ class ShardingFilterIterDataPipe(IterDataPipe):
     """
     def __init__(self, source_datapipe: IterDataPipe):
         self.source_datapipe = source_datapipe
+        self.sharding_groups: Dict[int, Any] = {}
         self.num_of_instances = 1
         self.instance_id = 0
 
     def is_shardable(self):
         return True
 
-    def apply_sharding(self, num_of_instances, instance_id):
-        self.num_of_instances = num_of_instances
-        self.instance_id = instance_id
+    def apply_sharding(self, num_of_instances, instance_id, sharding_group=None):
```

Comment:
I think this function has become more complicated with this change such that a doc string describing what it does would be a good idea, especially about what `sharding_group` does.

From what I can tell, the keys of `sharding_groups` can either be `None` or other numbers, and sharding is applied in the order of the group numbers.